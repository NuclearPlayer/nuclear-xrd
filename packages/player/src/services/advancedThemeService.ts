import { readTextFile } from '@tauri-apps/plugin-fs';

import {
  applyAdvancedTheme,
  parseAdvancedTheme,
  setThemeId,
} from '@nuclearplayer/themes';

import { setSetting, useSettingsStore } from '../stores/settingsStore';

export const loadAndApplyAdvancedThemeFromFile = async (
  path: string,
): Promise<void> => {
  const contents = await readTextFile(path);
  const json = JSON.parse(contents);
  const theme = parseAdvancedTheme(json);
  setThemeId('');
  applyAdvancedTheme(theme);
  await setSetting('core.theme.mode', 'advanced');
  await setSetting('core.theme.advanced.path', path);
};

export const applyAdvancedThemeFromSettingsIfAny = async (): Promise<void> => {
  const mode = useSettingsStore.getState().getValue('core.theme.mode');
  const path = useSettingsStore.getState().getValue('core.theme.advanced.path');
  if (mode === 'advanced' && typeof path === 'string' && path) {
    try {
      setThemeId('');
      await loadAndApplyAdvancedThemeFromFile(path);
    } catch {
      // ignore failures; user can re-pick a file in the UI
    }
  }
};
