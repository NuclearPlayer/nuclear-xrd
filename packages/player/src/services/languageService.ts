import { i18n } from '@nuclearplayer/i18n';

import { coreSettingsHost, getSetting } from '../stores/settingsStore';

export const changeLanguage = async (locale: string) => {
  await i18n.changeLanguage(locale);
};

export const applyLanguageFromSettings = async () => {
  const savedLanguage = getSetting('core.general.language');
  if (savedLanguage && typeof savedLanguage === 'string') {
    await changeLanguage(savedLanguage);
  }
};

export const initLanguageWatcher = () => {
  coreSettingsHost.subscribe('core.general.language', (value) => {
    if (value && typeof value === 'string') {
      void changeLanguage(value);
    }
  });
};
