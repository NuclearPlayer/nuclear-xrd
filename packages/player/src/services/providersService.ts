import type {
  ProviderDescriptor,
  ProviderKind,
  ProvidersHost,
} from '@nuclearplayer/plugin-sdk';

class ProvidersService implements ProvidersHost {
  #byKind = new Map<ProviderKind, Map<string, ProviderDescriptor>>();
  #byId = new Map<string, ProviderDescriptor>();

  register<T extends ProviderDescriptor>(provider: T): string {
    const kindMap = this.#byKind.get(provider.kind) ?? new Map();
    kindMap.set(provider.id, provider);
    this.#byKind.set(provider.kind, kindMap);
    this.#byId.set(provider.id, provider);
    return provider.id;
  }

  unregister(providerId: string): boolean {
    const current = this.#byId.get(providerId);
    if (!current) {
      return false;
    }
    this.#byId.delete(providerId);
    const kindMap = this.#byKind.get(current.kind);
    if (kindMap) {
      kindMap.delete(providerId);
      if (kindMap.size === 0) {
        this.#byKind.delete(current.kind);
      }
    }
    return true;
  }

  list<K extends ProviderKind = ProviderKind>(kind?: K) {
    if (kind) {
      const map = this.#byKind.get(kind as ProviderKind);
      return (map ? Array.from(map.values()) : []) as ProviderDescriptor<K>[];
    }
    const all: ProviderDescriptor[] = [];
    for (const map of this.#byKind.values()) {
      for (const value of map.values()) {
        all.push(value);
      }
    }
    return all as ProviderDescriptor<K>[];
  }

  get<T extends ProviderDescriptor>(providerId: string) {
    return this.#byId.get(providerId) as T | undefined;
  }

  clear() {
    this.#byKind.clear();
    this.#byId.clear();
  }
}

export const providersServiceHost: ProvidersHost = new ProvidersService();
