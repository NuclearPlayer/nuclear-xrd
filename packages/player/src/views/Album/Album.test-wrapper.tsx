import { RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SearchWrapper } from '../Search/Search.test-wrapper';

export const AlbumWrapper = {
  async mount(header: string): Promise<RenderResult> {
    const component = await SearchWrapper.mount('test album');
    const albums = await screen.findAllByTestId('card');
    if (albums.length === 0) {
      throw new Error('No albums found in search results');
    }
    await userEvent.click(albums[0]!);
    await screen.findByText(header);
    return component;
  },
  async mountNoWait(): Promise<RenderResult> {
    const component = await SearchWrapper.mount('test album');
    const albums = await screen.findAllByTestId('card');
    if (albums.length === 0) {
      throw new Error('No albums found in search results');
    }
    await userEvent.click(albums[0]!);
    await new Promise((r) => setTimeout(r, 0));
    return component;
  },
  getHeader: (name: string) => screen.getByRole('heading', { name }),
  getTracksTable: () => screen.queryByRole('table'),
};
