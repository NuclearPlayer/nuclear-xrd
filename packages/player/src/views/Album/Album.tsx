import { useParams } from '@tanstack/react-router';
import { FC } from 'react';

import { ScrollableArea } from '@nuclearplayer/ui';

import { AlbumHeader } from './components/AlbumHeader';
import { AlbumTrackList } from './components/AlbumTrackList';

type AlbumProps = Record<string, never>;

export const Album: FC<AlbumProps> = () => {
  const { providerId, albumId } = useParams({
    from: '/album/$providerId/$albumId',
  });

  return (
    <ScrollableArea className="bg-background">
      <AlbumHeader providerId={providerId} albumId={albumId} />
      <div className="p-6">
        <AlbumTrackList providerId={providerId} albumId={albumId} />
      </div>
    </ScrollableArea>
  );
};
