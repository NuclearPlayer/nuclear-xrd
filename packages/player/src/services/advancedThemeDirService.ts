import { join } from '@tauri-apps/api/path';
import {
  BaseDirectory,
  readDir,
  readTextFile,
  watch,
} from '@tauri-apps/plugin-fs';

import { parseAdvancedTheme } from '@nuclearplayer/themes';

import type { AdvancedThemeFile } from '../stores/advancedThemeStore';
import { useAdvancedThemeStore } from '../stores/advancedThemeStore';
import { useSettingsStore } from '../stores/settingsStore';
import { logFsError } from '../utils/logging';
import { ensureDir } from '../utils/path';
import { loadAndApplyAdvancedThemeFromFile } from './advancedThemeService';

let unwatch: (() => void) | null = null;

export const THEMES_DIR_NAME = 'themes';

export const ensureThemesDir = async (): Promise<string> => {
  return ensureDir(THEMES_DIR_NAME);
};

export const listAdvancedThemes = async (): Promise<AdvancedThemeFile[]> => {
  const dir = await ensureThemesDir();
  let entries: Array<{ name?: string; isDirectory?: boolean }> = [];
  try {
    entries = await readDir(dir, { baseDir: BaseDirectory.AppData });
  } catch (error) {
    await logFsError({
      scope: 'themes',
      command: 'fs.readDir',
      targetPath: dir,
      error,
      withToast: true,
      toastMessage: 'Failed to read themes directory',
    });
    return [];
  }
  const themes: AdvancedThemeFile[] = [];
  for (const e of entries) {
    if (e.isDirectory) {
      continue;
    }
    const name = e.name ?? '';
    if (!name.toLowerCase().endsWith('.json')) {
      continue;
    }
    const path = await join(dir, name);
    try {
      const text = await readTextFile(path, { baseDir: BaseDirectory.AppData });
      const json = JSON.parse(text);
      const parsed = parseAdvancedTheme(json);
      themes.push({ path, name: parsed.name });
    } catch (error) {
      await logFsError({
        scope: 'themes',
        command: 'fs.readTextFile',
        targetPath: path,
        error,
        withToast: true,
        toastMessage: 'Failed to read theme file',
      });
    }
  }
  themes.sort((a, b) => a.name.localeCompare(b.name));
  return themes;
};

export const refreshAdvancedThemeList = async (): Promise<void> => {
  const themes = await listAdvancedThemes();
  useAdvancedThemeStore.getState().setThemes(themes);
};

export const startAdvancedThemeWatcher = async (): Promise<void> => {
  const dir = await ensureThemesDir();
  await refreshAdvancedThemeList();
  if (unwatch) {
    return;
  }
  try {
    unwatch = await watch(
      dir,
      async (event) => {
        await refreshAdvancedThemeList();

        const state = useSettingsStore.getState();
        const mode = state.getValue('core.theme.mode');
        const currentPath = state.getValue('core.theme.advanced.path');

        if (
          mode !== 'advanced' ||
          typeof currentPath !== 'string' ||
          !currentPath
        ) {
          return;
        }

        if (!event.paths.some((p) => p.endsWith(currentPath))) {
          return;
        }

        try {
          await loadAndApplyAdvancedThemeFromFile(currentPath);
        } catch (error) {
          await logFsError({
            scope: 'themes',
            command: 'fs.watch',
            targetPath: currentPath,
            error,
            withToast: true,
            toastMessage: 'Theme reload failed',
          });
        }
      },
      { baseDir: BaseDirectory.AppData },
    );
  } catch (error) {
    await logFsError({
      scope: 'themes',
      command: 'fs.watch',
      targetPath: dir,
      error,
      withToast: true,
      toastMessage: "Couldn't watch themes directory",
    });
  }
};

export const stopAdvancedThemeWatcher = (): void => {
  if (unwatch) {
    unwatch();
    unwatch = null;
  }
};
