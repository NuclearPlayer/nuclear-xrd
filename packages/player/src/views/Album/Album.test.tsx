import { screen } from '@testing-library/react';

import { providersServiceHost } from '../../services/providersService';
import { MetadataProviderBuilder } from '../../test/builders/MetadataProviderBuilder';
import { AlbumWrapper } from './Album.test-wrapper';

describe('Album view', () => {
  beforeEach(() => {
    providersServiceHost.clear();
    const provider = new MetadataProviderBuilder()
      .withSearchCapabilities(['unified', 'albums'])
      .withAlbumMetadataCapabilities(['albumDetails'])
      .withSearch(async () => ({
        albums: [
          {
            title: 'Test Album',
            artists: [
              {
                name: 'Test Artist',
                source: {
                  provider: 'test-metadata-provider',
                  id: 'test-artist-id',
                },
              },
            ],
            artwork: {
              items: [
                {
                  url: 'https://img/album-cover.jpg',
                  purpose: 'cover',
                  width: 600,
                },
              ],
            },
            source: {
              provider: 'test-metadata-provider',
              id: 'test-album-id',
            },
          },
        ],
      }))
      .withFetchAlbumDetails(async () => ({
        title: 'Prism',
        artists: [
          {
            name: 'John Butler',
            roles: [],
          },
        ],
        tracks: [
          {
            title: 'Going Solo',
            artists: [
              {
                name: 'John Butler',
                source: {
                  provider: 'test-metadata-provider',
                  id: 'test-artist-id',
                },
              },
            ],
            source: {
              provider: 'test-metadata-provider',
              id: 'track-1',
            },
          },
          {
            title: 'King of California',
            artists: [
              {
                name: 'John Butler',
                source: {
                  provider: 'test-metadata-provider',
                  id: 'test-artist-id',
                },
              },
            ],
            source: {
              provider: 'test-metadata-provider',
              id: 'track-2',
            },
          },
        ],
        releaseDate: {
          precision: 'year',
          dateIso: '2025-01-01',
        },
        genres: ['Folk', 'World', 'Country'],
        artwork: {
          items: [
            {
              url: 'https://img/album-cover.jpg',
              purpose: 'cover',
              width: 600,
            },
          ],
        },
        source: {
          provider: 'test-metadata-provider',
          id: 'test-album-id',
        },
      }))
      .build();

    providersServiceHost.register(provider);
  });

  it('(Snapshot) renders the album view', async () => {
    const component = await AlbumWrapper.mount('Prism');
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('shows loading states for album details and tracks', async () => {
    providersServiceHost.clear();
    const delay = () => {
      return new Promise<never>(() => {});
    };
    const provider = new MetadataProviderBuilder()
      .withId('query-cache-busted')
      .withName('The provider that never resolves')
      .withSearchCapabilities(['unified', 'albums'])
      .withAlbumMetadataCapabilities(['albumDetails'])
      .withSearch(async () => ({
        albums: [
          {
            title: 'Test Album',
            artists: [
              {
                name: 'Test Artist',
                source: {
                  provider: 'test-metadata-provider',
                  id: 'test-artist-id',
                },
              },
            ],
            source: {
              provider: 'test-metadata-provider',
              id: 'test-album-id',
            },
          },
        ],
      }))
      .withFetchAlbumDetails(delay)
      .build();
    providersServiceHost.register(provider);

    await AlbumWrapper.mountNoWait();

    expect(
      await screen.findByTestId('album-header-loader'),
    ).toBeInTheDocument();
    expect(
      await screen.findByTestId('album-tracks-loader'),
    ).toBeInTheDocument();
  });
});
