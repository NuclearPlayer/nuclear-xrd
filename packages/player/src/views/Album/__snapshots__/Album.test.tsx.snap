// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Album view > (Snapshot) renders the album view 1`] = `
<DocumentFragment>
  <div
    class="grid h-screen w-screen grid-rows-[auto_1fr_auto] overflow-hidden"
  >
    <header
      class="bg-background-secondary border-border flex h-12 items-center border-b-2 px-4"
    >
      <svg
        class="inline-flex h-6 w-6"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          stroke="currentColor"
          stroke-width="8"
          transform="matrix(.96157 0 0 .95989 1.9231 1.9198)"
        >
          <rect
            height="96"
            rx="19.2"
            ry="19.2"
            width="96"
            x="2"
            y="2"
          />
          <path
            d="m30 23.6 20 26.4 20-26.4zm20 26.4 20 26.4 20-26.4zm0 0h-40l20 26.4z"
          />
        </g>
      </svg>
      <div>
        <input
          class="bg-background border-border ml-4 w-80 rounded-md border-2 px-3 py-1 outline-none"
          data-testid="search-box"
          placeholder="Search"
          value="test "
        />
      </div>
      <div
        class="flex-1"
      />
      <button
        aria-checked="false"
        aria-label="Toggle theme"
        class="border-border relative inline-flex h-6 w-11 cursor-pointer items-center rounded-full border-2 transition-all focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-none data-checked:bg-primary bg-white data-disabled:cursor-not-allowed data-disabled:opacity-50 group"
        data-headlessui-state=""
        id="headlessui-switch-:r0:"
        role="switch"
        tabindex="0"
        type="button"
      >
        <span
          class="border-border pointer-events-none flex h-4 w-4 items-center justify-center rounded-full border-2 bg-white ring-0 transition-transform [&>svg]:fill-white dark:[&>svg]:fill-black translate-x-1 group-data-checked:translate-x-5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-sun text-foreground"
            fill="none"
            height="12"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="4"
            />
            <path
              d="M12 2v2"
            />
            <path
              d="M12 20v2"
            />
            <path
              d="m4.93 4.93 1.41 1.41"
            />
            <path
              d="m17.66 17.66 1.41 1.41"
            />
            <path
              d="M2 12h2"
            />
            <path
              d="M20 12h2"
            />
            <path
              d="m6.34 17.66-1.41 1.41"
            />
            <path
              d="m19.07 4.93-1.41 1.41"
            />
          </svg>
        </span>
      </button>
    </header>
    <div
      class="bg-background-secondary relative grid h-full min-h-0 grid-cols-[auto_1fr_auto]"
    >
      <div
        animate="[object Object]"
        class="bg-background-secondary border-border relative flex flex-col p-2 border-r-2"
        transition="[object Object]"
      >
        <span
          class="mb-4 flex flex-row items-center justify-end"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 right-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-left"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-panel-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M9 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          animate="[object Object]"
          class="flex-1 overflow-hidden"
          exit="[object Object]"
          initial="[object Object]"
          transition="[object Object]"
        >
          <div
            class="flex flex-1 flex-col items-start justify-start"
          >
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-compass"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m16.24 7.76-1.804 5.411a2 2 0 0 1-1.265 1.265L7.76 16.24l1.804-5.411a2 2 0 0 1 1.265-1.265z"
                    />
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                  </svg>
                  Explore
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-settings"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M9.671 4.136a2.34 2.34 0 0 1 4.659 0 2.34 2.34 0 0 0 3.319 1.915 2.34 2.34 0 0 1 2.33 4.033 2.34 2.34 0 0 0 0 3.831 2.34 2.34 0 0 1-2.33 4.033 2.34 2.34 0 0 0-3.319 1.915 2.34 2.34 0 0 1-4.659 0 2.34 2.34 0 0 0-3.32-1.915 2.34 2.34 0 0 1-2.33-4.033 2.34 2.34 0 0 0 0-3.831A2.34 2.34 0 0 1 6.35 6.051a2.34 2.34 0 0 0 3.319-1.915"
                    />
                    <circle
                      cx="12"
                      cy="12"
                      r="3"
                    />
                  </svg>
                  Preferences
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-library"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m16 6 4 14"
                    />
                    <path
                      d="M12 6v14"
                    />
                    <path
                      d="M8 8v12"
                    />
                    <path
                      d="M4 4v16"
                    />
                  </svg>
                  Collection
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors right-0"
        />
      </div>
      <main
        class="bg-background-secondary overflow-auto"
        data-testid="player-workspace-main"
      >
        <div
          class="relative h-full w-full overflow-hidden"
        >
          <div
            animate="center"
            class="absolute inset-0 h-full w-full"
            exit="exit"
            initial="enter"
            transition="[object Object]"
          >
            <div
              class="relative h-full w-full bg-background"
            >
              <div
                class="scrollbar-hide flex h-full w-full flex-col overflow-auto pr-6 pb-6"
                style="scrollbar-width: none;"
              >
                <div
                  class="bg-primary border-border flex flex-col gap-6 border-b-2 p-8 md:flex-row"
                >
                  <img
                    alt="Prism"
                    class="border-border h-60 w-60 border-2 object-cover select-none"
                    src="https://img/album-cover.jpg"
                  />
                  <div
                    class="flex flex-1 flex-col gap-4"
                  >
                    <div
                      class="flex flex-col gap-2"
                    >
                      <h1
                        class="text-5xl font-bold"
                      >
                        Prism
                      </h1>
                      <div
                        class="text-text-secondary text-lg"
                      >
                        by John Butler
                      </div>
                    </div>
                    <div
                      class="text-text-secondary flex flex-col flex-wrap gap-2 text-sm"
                    >
                      <div>
                        <span
                          class="text-text-primary font-semibold"
                        >
                          Genre:
                        </span>
                         Folk, World, Country
                      </div>
                      <div>
                        <span
                          class="text-text-primary font-semibold"
                        >
                          Year:
                        </span>
                         2025
                      </div>
                      <div>
                        <span
                          class="text-text-primary font-semibold"
                        >
                          Tracks:
                        </span>
                         2
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="p-6"
                >
                  <div
                    class="relative flex max-h-full overflow-auto"
                  >
                    <table
                      class="border-border relative w-full border-2"
                      role="table"
                    >
                      <thead>
                        <tr
                          class="border-border bg-primary border-b-2"
                          role="row"
                        >
                          <th
                            class="px-2 text-left cursor-pointer"
                            role="columnheader"
                          >
                            <span
                              class="flex items-center"
                            >
                              Artist
                            </span>
                          </th>
                          <th
                            class="px-2 text-left cursor-pointer"
                            role="columnheader"
                          >
                            <span
                              class="flex items-center"
                            >
                              Title
                            </span>
                          </th>
                        </tr>
                      </thead>
                      <tbody />
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
      <div
        animate="[object Object]"
        class="bg-background-secondary border-border relative flex flex-col p-2 border-l-2"
        transition="[object Object]"
      >
        <span
          class="mb-4 flex flex-row items-center justify-start"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 left-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-right"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-panel-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M15 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          animate="[object Object]"
          class="flex-1 overflow-hidden"
          exit="[object Object]"
          initial="[object Object]"
          transition="[object Object]"
        />
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors left-0"
        />
      </div>
    </div>
    <footer
      class="bg-background-secondary border-border flex h-16 items-center border-t-2 px-4"
    >
      <div
        class="grid w-full grid-cols-[minmax(0,1fr)_auto_minmax(0,1fr)] items-center gap-4"
      >
        <div
          class="min-w-0"
        >
          <div
            class="flex min-w-0 items-center gap-3"
          >
            <div
              class="border-border bg-background-secondary size-12 shrink-0 overflow-hidden rounded-md border-2"
            >
              <img
                alt=""
                class="size-full object-cover select-none"
                src="https://picsum.photos/64"
              />
            </div>
            <div
              class="min-w-0"
            >
              <div
                class="text-foreground truncate text-sm font-bold"
              >
                Song Title
              </div>
              <div
                class="text-foreground-secondary truncate text-xs"
              >
                Artist Name
              </div>
            </div>
          </div>
        </div>
        <div
          class="justify-self-center"
        >
          <div
            class="flex items-center justify-center gap-2"
          >
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-shuffle"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m18 14 4 4-4 4"
                />
                <path
                  d="m18 2 4 4-4 4"
                />
                <path
                  d="M2 18h1.973a4 4 0 0 0 3.3-1.7l5.454-8.6a4 4 0 0 1 3.3-1.7H22"
                />
                <path
                  d="M2 6h1.972a4 4 0 0 1 3.6 2.2"
                />
                <path
                  d="M22 18h-6.041a4 4 0 0 1-3.3-1.8l-.359-.45"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-skip-back"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17.971 4.285A2 2 0 0 1 21 6v12a2 2 0 0 1-3.029 1.715l-9.997-5.998a2 2 0 0 1-.003-3.432z"
                />
                <path
                  d="M3 20V4"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-play"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 5a2 2 0 0 1 3.008-1.728l11.997 6.998a2 2 0 0 1 .003 3.458l-12 7A2 2 0 0 1 5 19z"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-skip-forward"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 4v16"
                />
                <path
                  d="M6.029 4.285A2 2 0 0 0 3 6v12a2 2 0 0 0 3.029 1.715l9.997-5.998a2 2 0 0 0 .003-3.432z"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-repeat"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m17 2 4 4-4 4"
                />
                <path
                  d="M3 11v-1a4 4 0 0 1 4-4h14"
                />
                <path
                  d="m7 22-4-4 4-4"
                />
                <path
                  d="M21 13v1a4 4 0 0 1-4 4H3"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="justify-self-end"
        >
          <div
            class="flex items-center gap-2"
          >
            <button
              class="inline-flex cursor-pointer items-center rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-volume2 lucide-volume-2"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
                />
                <path
                  d="M16 9a5 5 0 0 1 0 6"
                />
                <path
                  d="M19.364 18.364a9 9 0 0 0 0-12.728"
                />
              </svg>
            </button>
            <div
              class="w-24"
            >
              <div
                class="flex w-full flex-col gap-2"
              >
                <div
                  class="relative h-6 w-full overscroll-contain"
                >
                  <div
                    class="pointer-events-none absolute inset-x-0 top-1/2 -translate-y-1/2"
                  >
                    <div
                      class="border-border bg-background-input relative h-3 w-full rounded-md border-2"
                    >
                      <div
                        class="bg-primary absolute top-0 left-0 h-full rounded-l"
                        style="width: 75%;"
                      />
                      <div
                        class="border-border absolute top-1/2 -ml-2 h-5 w-5 -translate-y-1/2 rounded-full border-2 bg-white"
                        style="left: 75%;"
                      />
                    </div>
                  </div>
                  <input
                    aria-labelledby=":r1:-label"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="75"
                    aria-valuetext="75"
                    class="absolute inset-0 z-[1] h-6 w-full appearance-none opacity-0 cursor-pointer"
                    id=":r1:"
                    max="100"
                    min="0"
                    step="1"
                    type="range"
                    value="75"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </footer>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;
